src/BDCS/API/V0.hs

```diff
@@ -326,11 +333,123 @@ recipesInfo repoLock branch recipe_names = liftIO $ RWL.withRead (gitRepoLock re
+-- >                 {
+-- >                     "commit": "97d483e8dd0b178efca9a805e5fd8e722c48ac8e",
+-- >                     "time": "Wed,  1 Mar 2017 13:29:37 -0800",
+-- >                     "summary": "Recipe nfs-server saved"
+-- >                 },
+-- >                 {
+-- >                     "commit": "857e1740f983bf033345c3242204af0ed7b81f37",
+-- >                     "time": "Wed,  1 Mar 2017 09:28:53 -0800",
+-- >                     "summary": "Recipe nfs-server saved",
+-- >                     "revision" : 1
+-- >                 }
+-- >             ],
+-- >             "total": 2
+-- >         },
+-- >         {
+-- >             "name": "ruby",
+-- >             "changes": [
+-- >                 {
+-- >                     "commit": "4b84f072befc3f4debbe1348d6f4b166f7c83d78",
+-- >                     "time": "Wed,  1 Mar 2017 13:32:09 -0800",
+-- >                     "summary": "Recipe ruby saved"
+-- >                 },
+-- >                 {
+-- >                     "commit": "85999253c1790367a860a344ea622971b7e0a050",
+-- >                     "time": "Wed,  1 Mar 2017 13:31:19 -0800",
+-- >                     "summary": "Recipe ruby saved"
+-- >                 }
+-- >             ],
+-- >             "total": 2
+-- >         }
+-- >     ],
+-- >     "errors": [
+-- >         {
+-- >             "recipe": "a-missing-recipe",
+-- >             "msg": "Error retrieving a-missing-recipe"
+-- >         }
+-- >     ]
+-- >     "offset": 0,
+-- >     "limit": 20
+-- > }
+--
+recipesChanges :: GitLock -> T.Text -> String -> Maybe Int -> Maybe Int -> Handler RecipesChangesResponse
+recipesChanges repoLock branch recipe_names moffset mlimit = liftIO $ RWL.withRead (gitRepoLock repoLock) $ do
+    let recipe_name_list = map T.pack (argify [recipe_names])
+    let offset = fromMaybe 0 moffset
+    let limit  = fromMaybe 20 mlimit
+    (changes, errors) <- allRecipeChanges recipe_name_list [] []
+    return $ RecipesChangesResponse changes errors offset limit
+  where
+    allRecipeChanges [] _ _ = return ([], [])
+    allRecipeChanges [recipe_name] changes_list errors_list =
+                     oneRecipeChange recipe_name changes_list errors_list
+    allRecipeChanges (recipe_name:xs) changes_list errors_list = do
+                     (new_changes, new_errors) <- oneRecipeChange recipe_name changes_list errors_list
+                     allRecipeChanges xs new_changes new_errors
+
+    oneRecipeChange recipe_name changes_list errors_list = do
+        result <- catch_recipe_changes recipe_name
+        return (new_changes result, new_errors result)
+      where
+        new_changes result = case result of
+            Left  _       -> changes_list
+            Right changes -> RecipeChanges recipe_name changes (length changes):changes_list
+
+        new_errors result = case result of
+            Left  err -> RecipesAPIError recipe_name (T.pack err):errors_list
+            Right _   -> errors_list
+
```

Comment:
In general, I don't really like case statements in Haskell unless absolutely necessary (like, there's a lot of branches).  For doing stuff on Either and Maybe especially, I like using multiple definitions for the same function.  So I'd probably rewrite these like:
```
new_changes (Right changes) -> RecipeChanges recipe_name changes (length changes):changes_list
new_changes _ -> changes_list

new_errors (Left err) -> RecipesAPIError recipe_name (T.pack err):errors_list
new_errors _ -> errors_list
```

Reply:
Oh, yeah! I like that better.